[SPARK-30338][SQL] Avoid unnecessary InternalRow copies in ParquetRowConverter

JoshRosen opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR modifies `ParquetRowConverter` to remove unnecessary `InternalRow.copy()` calls for structs that are directly nested in other structs. 

### Why are the changes needed?

These changes  can significantly improve performance when reading Parquet files that contain deeply-nested structs with many fields.

The `ParquetRowConverter` uses per-field `Converter`s for handling individual fields. Internally, these converters may have mutable state and may return mutable objects. In most cases, each `converter` is only invoked once per Parquet record (this is true for top-level fields, for example). However, arrays and maps may call their child element converters multiple times per Parquet record: in these cases we must be careful to copy any mutable outputs returned by child converters.

In the existing code, `InternalRow`s are copied whenever they are stored into _any_ parent container (not just maps and arrays). This copying can be especially expensive for deeply-nested fields, since a deep copy is performed at every level of nesting.

This PR modifies the code to avoid copies for structs that are directly nested in structs; see inline code comments for an argument for why this is safe. 

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

**Correctness**:  I added new test cases to `ParquetIOSuite` to increase coverage of nested structs, including structs nested in arrays: previously this suite didn't test that case, so we used to lack mutation coverage of this `copy()` code (the suite's tests still passed if I incorrectly removed the `.copy()` in all cases). I also added a test for maps with struct keys and modified the existing "map with struct values" test case include maps with two elements (since the incorrect omission of a `copy()` can only be detected if the map has multiple elements).

**Performance**: I put together a simple local benchmark demonstrating the performance problems:

First, construct a nested schema:

```scala
  case class Inner(
    f1: Int,
    f2: Long,
    f3: String,
    f4: Int,
    f5: Long,
    f6: String,
    f7: Int,
    f8: Long,
    f9: String,
    f10: Int
  )
  
  case class Wrapper1(inner: Inner)
  case class Wrapper2(wrapper1: Wrapper1)
  case class Wrapper3(wrapper2: Wrapper2)
```

`Wrapper3`'s schema looks like:

```
root
 |-- wrapper2: struct (nullable = true)
 |    |-- wrapper1: struct (nullable = true)
 |    |    |-- inner: struct (nullable = true)
 |    |    |    |-- f1: integer (nullable = true)
 |    |    |    |-- f2: long (nullable = true)
 |    |    |    |-- f3: string (nullable = true)
 |    |    |    |-- f4: integer (nullable = true)
 |    |    |    |-- f5: long (nullable = true)
 |    |    |    |-- f6: string (nullable = true)
 |    |    |    |-- f7: integer (nullable = true)
 |    |    |    |-- f8: long (nullable = true)
 |    |    |    |-- f9: string (nullable = true)
 |    |    |    |-- f10: integer (nullable = true)
```

Next, generate some fake data:

```scala
  val data = spark.range(1, 1000 * 1000 * 25, 1, 1).map { i =>
    Wrapper3(Wrapper2(Wrapper1(Inner(
      i.toInt,
      i * 2,
      (i * 3).toString,
      (i * 4).toInt,
      i * 5,
      (i * 6).toString,
      (i * 7).toInt,
      i * 8,
      (i * 9).toString,
      (i * 10).toInt
    ))))
  }

  data.write.mode("overwrite").parquet("/tmp/parquet-test")
```

I then ran a simple benchmark consisting of

```
spark.read.parquet("/tmp/parquet-test").selectExpr("hash(*)").rdd.count()
```

where the `hash(*)` is designed to force decoding of all Parquet fields but avoids `RowEncoder` costs in the `.rdd.count()` stage.

In the old code, expensive copying takes place at every level of nesting; this is apparent in the following flame graph:

![image](https://user-images.githubusercontent.com/50748/71389014-88a15380-25af-11ea-9537-3e87a2aef179.png)

After this PR's changes, the above toy benchmark runs ~30% faster.

joshrosen-stripe committed:
Add new "array of struct" test case to ParquetIOSuite

JoshRosen committed:
Avoid unnecessary copy of nested internal rows.

JoshRosen committed:
Add comments.

SparkQA commented:
**[Test build #115677 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115677/testReport)** for PR 26993 at commit [`2ed8ea9`](https://github.com/apache/spark/commit/2ed8ea96d0ba4e03f83c813327fbd6b86e72dc4b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20474/
Test PASSed.

joshrosen-stripe commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:
```diff
@@ -318,10 +318,31 @@ private[parquet] class ParquetRowConverter(
+            //      we don't need to copy because copying will be done in the final
+            //      UnsafeProjection, or
+            //   2. The path from the schema root to this field contains a map or array,
+            //      in which case we will perform a recursive defensive copy via the
```
Correctness relies on the copy actually being a deep copy. Looking elsewhere in this file, we have comments like

```
    // NOTE: We can't reuse the mutable Map here and must instantiate a new `Map` for the next
    // value.  `Row.copy()` only copies row cells, it doesn't do deep copy to objects stored in row
    // cells.
```

which suggest that certain copying might be shallow, so it's important to double-check and make sure that the copies are indeed deep.

Here, the state being copied is an `InternalRow`. To be more specific, it's actually a `SpecificInternalRow` (I'll update the `.asInstanceOf` cast below to reflect this). `SpecificInternalRow` extends `BaseGenericInternalRow` and #18483 changed that to implement a deep-copy, recursively copying maps, arrays, and structs.

joshrosen-stripe replied:
I think the existing comment about `Row.copy()` is outdated, so we might be able to optimize those other parts of the code, too; I'm going to defer that to future work / another PR, though.

JoshRosen replied:
**Update**: in #27089 I'm removing these other unnecessary ArrayBuffer copies.

JoshRosen committed:
More specific cast; add braces.

SparkQA commented:
**[Test build #115680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115680/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20477/
Test PASSed.

SparkQA commented:
**[Test build #115677 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115677/testReport)** for PR 26993 at commit [`2ed8ea9`](https://github.com/apache/spark/commit/2ed8ea96d0ba4e03f83c813327fbd6b86e72dc4b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115677/
Test FAILed.

SparkQA commented:
**[Test build #115680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115680/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115680/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #115702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115702/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20497/
Test PASSed.

SparkQA commented:
**[Test build #115702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115702/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115702/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20520/
Test PASSed.

SparkQA commented:
**[Test build #115727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115727/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).

SparkQA commented:
**[Test build #115727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115727/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115727/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #115742 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115742/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20536/
Test PASSed.

SparkQA commented:
**[Test build #115742 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115742/testReport)** for PR 26993 at commit [`3fb3391`](https://github.com/apache/spark/commit/3fb3391edcee6c467b0189de2dcdaf4dac610083).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115742/
Test PASSed.

JoshRosen committed:
Update tests to match existing style; strengthen map test

JoshRosen committed:
Add test for struct as map key type

SparkQA commented:
**[Test build #115874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115874/testReport)** for PR 26993 at commit [`fffe72b`](https://github.com/apache/spark/commit/fffe72b435c8f9f2d4cd0a0742fd7bfd247eabf7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20666/
Test PASSed.

SparkQA commented:
**[Test build #115874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115874/testReport)** for PR 26993 at commit [`fffe72b`](https://github.com/apache/spark/commit/fffe72b435c8f9f2d4cd0a0742fd7bfd247eabf7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115874/
Test PASSed.

JoshRosen commented:
I've removed the `[WIP]` tag and I think this is now ready for review.

 I've updated the existing "map with struct values" test so that it uses maps with multiple values (previously, we only tested with maps that contained one entry and that's insufficient to detect struct-copying problems (i.e. the old test would still pass if I completely removed the `.copy()` calls, whereas the new test will fail). For completeness, I also added a test for "maps with struct _keys_". I also changed the style of the new "struct of structs" test case in order to better match the style of this suite's existing tests.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:
```diff
@@ -318,10 +318,32 @@ private[parquet] class ParquetRowConverter(
         new ParquetMapConverter(parquetType.asGroupType(), t, updater)
 
       case t: StructType =>
+        val wrappedUpdater = {
```
@JoshRosen, no big deal at all but how about we put the JIRA ID somewhere in the comment?

JoshRosen replied:
Good idea: I added a JIRA reference in https://github.com/apache/spark/pull/26993/commits/e6945e88a24d51551cba105b5e7e3825bc5e0a69

HyukjinKwon commented:
I happened to take a cursory look and seems pretty fine.

JoshRosen committed:
Add JIRA number to comment

SparkQA commented:
**[Test build #115981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115981/testReport)** for PR 26993 at commit [`e6945e8`](https://github.com/apache/spark/commit/e6945e88a24d51551cba105b5e7e3825bc5e0a69).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20773/
Test PASSed.

SparkQA commented:
**[Test build #115981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115981/testReport)** for PR 26993 at commit [`e6945e8`](https://github.com/apache/spark/commit/e6945e88a24d51551cba105b5e7e3825bc5e0a69).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115981/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #116000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116000/testReport)** for PR 26993 at commit [`e6945e8`](https://github.com/apache/spark/commit/e6945e88a24d51551cba105b5e7e3825bc5e0a69).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20792/
Test PASSed.

SparkQA commented:
**[Test build #116000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116000/testReport)** for PR 26993 at commit [`e6945e8`](https://github.com/apache/spark/commit/e6945e88a24d51551cba105b5e7e3825bc5e0a69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116000/
Test PASSed.

joshrosen-stripe commented:
I ran a "read and hash all columns" benchmark on datasets with real-world schemas; these schemas contained ~50-300+ fields at various depths of nesting. The benchmark code looked roughly like

```scala
val data = spark.read.parquet(args.list("input"): _*)
data.select(hash($"*").as("hash")).groupBy().sum("hash").collect()
```

Comparing the map/scan stages' `Total Time Across All Tasks` metrics (from the Spark UI's "Stage Details" pages), it looks like this patch's changes result in ~6-25% time savings for this benchmark.

joshrosen-stripe commented:
@cloud-fan @dongjoon-hyun @viirya, could you take a look at this PR optimizing nested struct handling in `ParquetRowConverter`? I'm tagging this group because it looks like you've all helped to review recent changes to this file.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:
```diff
@@ -318,10 +318,33 @@ private[parquet] class ParquetRowConverter(
+            // `updater` is a RowUpdater, implying that the parent container is a struct.
+            // We do NOT need to perform defensive copying here because either:
+            //
+            //   1. The path from the schema root to this field consists only of nested
```
When we have deeply nested struct inside an array, is it the first case here?

I think it is fine because at the element converter the top level struct inside an array element will do the defensive copying. So in nested struct converter, we will see RowUpdater from parent struct so don't need defensive copying too.

Just maybe good to also update it in the doc.

joshrosen-stripe replied:
Yes, that's right. After thinking about this some more, I think I've come up with a clearer explanation and have updated the code comment: https://github.com/apache/spark/pull/26993/commits/4651b2fd724a56515c087903284682c9ba947c31

joshrosen-stripe committed:
Re-word code comment based on review feedback.

SparkQA commented:
**[Test build #116188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116188/testReport)** for PR 26993 at commit [`4651b2f`](https://github.com/apache/spark/commit/4651b2fd724a56515c087903284682c9ba947c31).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20981/
Test PASSed.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:
```diff
@@ -204,6 +204,23 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
     }
   }
 
+  testStandardAndLegacyModes("array of struct") {
```
Do we have a test for array of struct of struct?

joshrosen-stripe replied:
I added a new test case for this in https://github.com/apache/spark/pull/26993/commits/0f1af94c6cfc72651c3209bc0d31ae1bef140988

viirya reviewed (commented):
Looks correct and pretty good for performance improvement.

joshrosen-stripe committed:
Add test case for array of nested struct

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20985/
Test PASSed.

SparkQA commented:
**[Test build #116193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116193/testReport)** for PR 26993 at commit [`0f1af94`](https://github.com/apache/spark/commit/0f1af94c6cfc72651c3209bc0d31ae1bef140988).

SparkQA commented:
**[Test build #116188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116188/testReport)** for PR 26993 at commit [`4651b2f`](https://github.com/apache/spark/commit/4651b2fd724a56515c087903284682c9ba947c31).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116188/
Test FAILed.

SparkQA commented:
**[Test build #116193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116193/testReport)** for PR 26993 at commit [`0f1af94`](https://github.com/apache/spark/commit/0f1af94c6cfc72651c3209bc0d31ae1bef140988).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116193/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
